ui/app/js/controllers/required_signoffs_modal_base_controller.js

```diff
@@ -0,0 +1,349 @@
+                      return; // exit forEach
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+
+        first = false;
+
+        var addScheduledChangeCallback = function(data, deferred) {
+          return function(response) {
+            if ($scope.mode === "channel") {
+              if (data["change_type"] === "insert") {
+                required_signoffs[$scope.product]["channels"][$scope.channel][data["role"]] = {
+                  "signoffs_required": 0,
+                  "data_version": null,
+                  "sc": {
+                    // TODO: We should really be setting this, but the backend doesn't
+                    // return them.
+                    "required_signoffs": {},
+                    "signoffs_required": data["signoffs_required"] || 0,
+                    "sc_id": response["sc_id"],
+                    "scheduled_by": current_user,
+                    "sc_data_version": 1,
+                    "signoffs": {},
+                    "change_type": data["change_type"],
+                  },
+                };
+              }
+              else {
+                required_signoffs[$scope.product]["channels"][$scope.channel][data["role"]]["sc"] = {
+                  "required_signoffs": {},
+                  "signoffs_required": data["signoffs_required"] || 0,
+                  "sc_id": response["sc_id"],
+                  "scheduled_by": current_user,
+                  "sc_data_version": 1,
+                  "signoffs": {},
+                  "change_type": data["change_type"],
+                };
+              }
+            }
+            else {
+              if (data["change_type"] === "insert") {
+                required_signoffs[$scope.product]["permissions"][data["role"]] = {
+                  "signoffs_required": 0,
+                  "data_version": null,
+                  "sc": {
+                    "required_signoffs": {},
+                    "signoffs_required": data["signoffs_required"] || 0,
+                    "sc_id": response["sc_id"],
+                    "scheduled_by": current_user,
+                    "sc_data_version": 1,
+                    "signoffs": {},
+                    "change_type": data["change_type"],
+                  },
+                };
+              }
+              else {
+                required_signoffs[$scope.product]["permissions"][data["role"]]["sc"] = {
+                  "required_signoffs": {},
+                  "signoffs_required": data["signoffs_required"] || 0,
+                  "sc_id": response["sc_id"],
+                  "scheduled_by": current_user,
+                  "sc_data_version": 1,
+                  "signoffs": {},
+                  "change_type": data["change_type"],
+                };
+              }
+            }
+            deferred.resolve();
+          };
+        };
+        var errorCallback = function(data, deferred) {
```

Comment:
I'm not sure. Most of what this does is manipulate things from the controller ($scope variables, and deferred resolutions), so it may be weird to do it elsewhere...